### ISSUE TITLE
Endpoints - Exporting to CSV results in IndexError: list index out of range or KeyError: 'Study Instance UID'

### ISSUE THREAD

**Issue by DVHmonkey** [cutright/DVH-Analytics#20]
Happy holidays, and thanks for all the work.

Doing a test run with 57 cases, and attempting to export Endpoints (with 2 added endpoints) to CSV resulted in an IndexError: list index out of range error. Terminal output is below.

![IndexError](https://user-images.githubusercontent.com/58673531/71496616-8254f500-2897-11ea-837d-ee53eeddd6ec.jpg)

With only 1 added endpoint, I get KeyError: 'Study Instance UID'.

![StudyInstanceUID](https://user-images.githubusercontent.com/58673531/71496646-ae707600-2897-11ea-97f0-1357930e0bfc.jpg)

Thanks and regards.

**Comment by cutright** [comment]
This appears to be two separate issues.  I can reproduce the 2nd error simply by attempting a 2nd export without making any changes to endpoints.  I've since added a deepcopy() as shown below to fix this issue.  The current master now has this edit.

https://github.com/cutright/DVH-Analytics/blob/a147162221a8a22921c05d0478e0b03881105547/dvha/models/data_table.py#L321-L328

However, I'm not sure I understand the first error yet and I can't reproduce it.  I'll poke around some more.

As a side note, pasting terminal output as text would be easier to read (and search).  Yellow on white is pretty rough on the eyes.

**Comment by cutright** [comment]
Are you still having issues?

If so, could you give the 0.6.4dev branch a try?

**Comment by DVHmonkey** [comment]
Thanks, and happy new year. I just got back into work today, will try it as soon as I can.

**Comment by DVHmonkey** [comment]
Have tried it out with v0.64dev, with the following terminal output.

Traceback (most recent call last):

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-0.6.4dev\DVH-Analytics-0.6.4dev\dvha\models\endpoint.py", line 233, in on_export_csv
    save_data_to_file(self.parent, "Export Endpoints to CSV", self.get_csv())

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-0.6.4dev\DVH-Analytics-0.6.4dev\dvha\models\endpoint.py", line 224, in get_csv
    csv = self.data_table[1].get_csv(extra_column_data=uid)

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-0.6.4dev\DVH-Analytics-0.6.4dev\dvha\models\data_table.py", line 327, in get_csv
    self.insert_columns_into_data_for_csv(data, extra_column_data)

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-0.6.4dev\DVH-Analytics-0.6.4dev\dvha\models\data_table.py", line 380, in insert_columns_into_data_for_csv
    self.insert_column_into_data_for_csv(data_for_csv, columns_dict[index], index)

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-0.6.4dev\DVH-Analytics-0.6.4dev\dvha\models\data_table.py", line 367, in insert_column_into_data_for_csv
    row.insert(index, columns_dict_value['data'][i])

IndexError: list index out of range


I found something strange. While trying different permutations of query/search and numerical filter settings, I found that the export sometimes works under certain conditions. For context, the studies in my current project are divided into 3 different fractionation schedules: 78GyE/39fr, 74Gy/37fr and 70GyE/28fr. Filtering by ROI Max Dose, e.g. (65-73Gy, 76-79Gy, etc) sometimes gives me a set of studies (160 studies thereabouts) that I can export with up to 10 endpoints (max that I've tried so far). Filtering by Rx Dose doesn't, for some reason.

Perhaps a few cases in my pool are somehow screwing up the export function; will experiment further.

Thanks and regards.

**Comment by cutright** [comment]
Can you comment out lines 326 and 327 in dvha/models/data_table.py?  I'm wondering if the issue is related to trying to add the study instance uid to the csv data.

```
if extra_column_data:
    self.insert_columns_into_data_for_csv(data, extra_column_data)
```

Or maybe there are some values that contain commas somehow?  I thought I addressed that everywhere, but maybe not?

Thanks for your help... and patience.

**Comment by DVHmonkey** [comment]
Thanks for your reply. Sorry for not replying earlier, as I was on medical leave for the past couple of of days.

The lines are below:

326        if extra_column_data:
327            self.insert_columns_into_data_for_csv(data, extra_column_data)

I'll try and see if there any strange values in my case data.

Thanks, and regards.

**Comment by cutright** [comment]
No worries.

My bet is that there is an MRN, UID, or ROI name with a comma.  It looks like I don't account for that in `DataTable.get_csv()`.

**Comment by cutright** [comment]
Oh, so I do account for commas in `DataTable.data_for_csv`, but not in the `extra_column_data`.  In the case of exporting endpoints, this refers to StudyInstanceUID.

Perhaps at least one of your imported plans has a StudyInstanceUID with a comma somehow?  If so, were they manually changed?  If they weren't changed but do contain commas, which planning system and version?

If it's not StudyInstanceUID, I'm at a loss.

https://github.com/cutright/DVH-Analytics/blob/b532ff192a30a2c4e05d747df1645b1acdd35063/dvha/models/data_table.py#L313-L352

https://github.com/cutright/DVH-Analytics/blob/b532ff192a30a2c4e05d747df1645b1acdd35063/dvha/models/endpoint.py#L227-L236

**Comment by cutright** [comment]
Can you give the 0.6.6dev branch a shot?

Edited the following in `DataTable.insert_column_into_data_for_csv()`
In v0.6.5:
https://github.com/cutright/DVH-Analytics/blob/b532ff192a30a2c4e05d747df1645b1acdd35063/dvha/models/data_table.py#L366-L367

Changed to the following in v0.6.6dev:
https://github.com/cutright/DVH-Analytics/blob/83ac78d2bb24fab27867a5365319fbc3ab6a31dc/dvha/models/data_table.py#L366-L367

**Comment by cutright** [comment]
Correction, try the current master, 0.6.6 already went live.

**Comment by DVHmonkey** [comment]
Thanks for your reply. I just tried the current master with 400 cases, 2 ROIs and 3 endpoints and unfortunately was unable to export. Terminal output is below:

Traceback (most recent call last):

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-master2020-01-04\DVH-Analytics-master\dvha\models\endpoint.py", line 201, in on_export_csv
    save_data_to_file(self.parent, "Export Endpoints to CSV", self.get_csv())

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-master2020-01-04\DVH-Analytics-master\dvha\models\endpoint.py", line 198, in get_csv
    return self.data_table.get_csv(extra_column_data=uid)

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-master2020-01-04\DVH-Analytics-master\dvha\models\data_table.py", line 323, in get_csv
    self.insert_columns_into_data_for_csv(data, extra_column_data)

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-master2020-01-04\DVH-Analytics-master\dvha\models\data_table.py", line 376, in insert_columns_into_data_for_csv
    self.insert_column_into_data_for_csv(data_for_csv, columns_dict[index], index)

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-master2020-01-04\DVH-Analytics-master\dvha\models\data_table.py", line 363, in insert_column_into_data_for_csv
    row.insert(index, columns_dict_value['data'][i])

IndexError: list index out of range

Have been out of the office for the past few days, will continue looking through the data from my cases to see if anything weird is mixed in there.

Thanks and regards.

**Comment by cutright** [comment]
400 cases! SWEET!

Can you verify your version? Version 0.6.6 went live on Jan 9, but your output refers to DVH-Analytics-master2020-01-04.

**Comment by DVHmonkey** [comment]
D'oh, you're right, it was an old version. Running the latest master v0.6.7rev1 with same queries results in an error as well; terminal output below.

Traceback (most recent call last):

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-master 2020-01-14\DVH-Analytics-master\dvha\models\endpoint.py", line 239, in on_export_csv
    save_data_to_file(self.parent, "Export Endpoints to CSV", self.get_csv())

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-master 2020-01-14\DVH-Analytics-master\dvha\models\endpoint.py", line 230, in get_csv
    csv = self.data_table[1].get_csv(extra_column_data=uid)

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-master 2020-01-14\DVH-Analytics-master\dvha\models\data_table.py", line 327, in get_csv
    self.insert_columns_into_data_for_csv(data, extra_column_data)

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-master 2020-01-14\DVH-Analytics-master\dvha\models\data_table.py", line 380, in insert_columns_into_data_for_csv
    self.insert_column_into_data_for_csv(data_for_csv, columns_dict[index], index)

  File "C:\Users\MRI\Desktop\Temporary Statistics Software\DVH-Analytics-master 2020-01-14\DVH-Analytics-master\dvha\models\data_table.py", line 367, in insert_column_into_data_for_csv
    row.insert(index, str(columns_dict_value['data'][i]).replace(',', ';'))

IndexError: list index out of range

On a side note: I am working on eventually getting the entire case pool at this facility into a database, just for kicks. Currently in the process of getting the proprietary file formats into DICOM so I can import them. Due to several workarounds that had to be used, I can't use some functions of DVHa - so for the moment I am using DVHa as an easy way to query multiple endpoints of a large amount of cases at the same time. Might eventually end up at a different facility with more compatible DICOMs, so I'm looking forward to seeing future progress.

**Comment by cutright** [comment]
The difficulty is that at this point, I'd have to iteratively insert print statements into the source code, while using your data, to figure out what's happening.

For example, I'd change:
https://github.com/cutright/DVH-Analytics/blob/88b96241bb8058ec4d3c4596f648679977a7459b/dvha/models/data_table.py#L365-L367

to:
```
columns_dict_value['data'].insert(0, columns_dict_value['title'])
print('data_for_csv', data_for_csv)
print('columns_dict_value', columns_dict_value['data'])
print('data_for_csv length:', len(data_for_csv))
print('columns_dict_value length:', len(columns_dict_value['data']))
for i, row in enumerate(data_for_csv):
    row.insert(index, str(columns_dict_value['data'][i]).replace(',', ';'))
```

Based on the IndexError, it looks looks like `data_for_csv` has more rows than does `columns_dict_value['data']`.  I can't discern why from the information given.

I suspect the underlying issue is related to your DICOM conversion process.  Perhaps the format of some of your DICOM tags are different from the TPS vendors.

**Comment by DVHmonkey** [comment]
Discovered a possible cause. The error only occurs when I have queried two or more different ROI Names. My current use case requires endpoints of two different structures, so if I query the database for both ROI Names, call up endpoints and attempt to export to csv, it results in the errors and terminal output described in my previous replies.

Limiting the query to 1 ROI Name allowed me to export to csv without any problems, so I just did it once for each ROI Name.

In conclusion, my error merely boils down to "Can't export Endpoints to CSV when more than one ROI Name is queried", and I worked around it as above.

Thanks, and regards.

**Comment by cutright** [comment]
Good work!

Turns out it's not explicitly querying two ROI names.  My first attempt to reproduce worked just fine. Luckily my second query crashed.  Will follow-up shortly.

**Comment by cutright** [comment]
The issue was in `dvha.models.dvh.DVH`. The class stored the input StudyInstanceUIDs rather than the returned values. Let's just say UID hereafter.

I send a unique list of UIDs which ensures the SQL return only includes results that have a UID in this list, but you may get multiple returns for one UID or a UID may not be in the return at all.

So it turns out the the number of rows in the endpoint table often didn't match the number of stored UIDs regardless of the number of roi names in the query.  But if there were at least as many UIDs as Endpoint rows, DVHA would not crash.

This impacts the global export feature as well, for the DVH summary and DVHs themselves.

@DVHmonkey Can you give the 0.6.9dev branch a try?  Fairly certain your issue is resolved in that branch.

**Comment by DVHmonkey** [comment]
Brilliant! Just tested it with v0.6.9dev, works perfectly. Just for kicks, tried it with all 400 cases, 5 ROIs each and 15 endpoints. Exported to csv flawlessly.

Thanks!

**Comment by cutright** [comment]
Just released v0.6.9.  Issue resolved.
